### PR TITLE
[feat] add speculative decoding.

### PR DESCRIPTION
**Pull request by liutongxuan**

### CODE REVIEW

File: src/engine/engine.cpp [hunk 1]
```diff
@@ -303,4 +303,39 @@ OutputParameters Engine::execute_model(const std::vector<Sequence*>& batch) {
...
+                                 &input_params,
+                                 &sampling_params);
+  if (workers_.size() == 1) {
+    auto output = workers_[0]->validate(
+        flatten_token_ids, flatten_positions, input_params, sampling_params);
+    output.index_select(seq_idxes);
+    return output;
+  }
+
+  std::vector<folly::SemiFuture<OutputParameters>> futures;
```
Comment (guocuimi) [inline src/engine/engine.cpp:328]: nit: in my understanding, all workers should return same logits from last step if they are initialized with same seed. If it is the case, sounds validattion can be done only on one worker.

File: src/engine/utils.cpp [hunk 2]
```diff
@@ -208,4 +208,116 @@ void Utils::prepare_inputs(const std::vector<Sequence*>& batch,
...
 }
 
+void Utils::prepare_validate_inputs(const std::vector<Sequence*>& batch,
+                                    int32_t block_size,
+                                    torch::Tensor* flatten_token_ids,
+                                    torch::Tensor* flatten_positions,
+                                    torch::Tensor* seq_idxes,
+                                    InputParameters* input_params,
+                                    SamplingParameters* sampling_params) {
+  std::vector<int32_t> flatten_tokens_vec;
```
Comment (guocuimi) [inline src/engine/utils.cpp:218]: Just a heads up, this part may need to change after attention kernel change.

File: src/scheduler/speculative_scheduler.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  Engine* llm_engine_;
+  Engine* ssm_engine_;
+
+  BlockManager* llm_block_manager_;
+  BlockManager* ssm_block_manager_; 
+
+  std::unique_ptr<Tokenizer> tokenizer_;
+
+  SchedulerPolicy* scheduler_policy_;
```
Comment (guocuimi) [inline src/scheduler/speculative_scheduler.h:43]: use std::unique_ptr instead to avoid call new/delete explicitly

File: src/scheduler/speculative_scheduler.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  llm_block_manager_ = llm_engine_->block_manager();
+  GCHECK(ssm_engine_ != nullptr);
+  ssm_block_manager_ = ssm_engine_->block_manager();
+
+  tokenizer_ = llm_engine_->tokenizer();
+  GCHECK(llm_block_manager_ != nullptr);
+  GCHECK(ssm_block_manager_ != nullptr);
+  GCHECK(tokenizer_ != nullptr);
+
+  response_handler_ = new ResponseHandler(llm_block_manager_,
```
Comment (guocuimi) [inline src/scheduler/speculative_scheduler.cpp:29]: use std::unique_ptr instead

File: src/scheduler/scheduler_config.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  static SchedulerPolicyType FCFS;
+  static SchedulerPolicyType PSA;
+
+ private:
+  std::string type_;
+};
+
+SchedulerPolicyType SchedulerPolicyType::FCFS("fcfs");
+SchedulerPolicyType SchedulerPolicyType::PSA("psa");
```
Comment (guocuimi) [inline src/scheduler/scheduler_config.h:34]: what is psa?

File: src/scheduler/speculative_scheduler.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                                          tokenizer_.get());
+  scheduler_policy_ = new FCFSSchedulerPolicy(response_handler_,
+                                              llm_block_manager_);
```
Comment (guocuimi) [inline src/scheduler/speculative_scheduler.cpp:32]: nit: sounds you need to construct a joint block manager to include both slm and llm, so that both small and large model have a same memory allocation.

File: src/scheduler/speculative_scheduler.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  // run validation on llm.
+  auto output_parameters = validate(spec_sequences_batch);
+
+  const auto& next_tokens = output_parameters.next_tokens;
+  const int64_t num_seqs = next_tokens.sizes()[0];
+  GCHECK(num_seqs == spec_sequences_batch.size());
+
+  for (int64_t i = 0; i < num_seqs; ++i) {
+    Sequence* seq = spec_sequences_batch[i];
+    auto seq_next_tokens = next_tokens.index_select(0,
```
Comment (guocuimi) [inline src/scheduler/speculative_scheduler.cpp:74]: for pytorch c++ interface, it is not easy to read without parameter name. i would suggest add parameter name in comments: for example: index_select(/*dim=*/0, /*...*/)

File: src/scheduler/speculative_scheduler.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        torch::tensor(i, torch::kInt));
```
Comment (guocuimi) [inline src/scheduler/speculative_scheduler.cpp:75]: in my understanding, the tensor type needs to be Int64, otherwise, it triggers a core dump.

File: src/scheduler/speculative_scheduler.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    const int64_t* new_token_ids = next_tokens.data_ptr<int64_t>();
+    for (int64_t i = 0; i < num_seqs; ++i) {
+      auto seq = spec_batch[i];
+      const auto next_token_id = static_cast<int32_t>(new_token_ids[i]);
+      seq->append_new_token_id(next_token_id);
+      // record speculative token ids
+      seq->append_spec_token_id(next_token_id);
+
+      if (!seq->is_finished()) {
+        next_spec_batch.emplace_back(seq);
```
[the remark this answers is not in the document]
Comment (liutongxuan) [reply src/scheduler/speculative_scheduler.cpp:109]: yes, update_valid_token_ids is used to update the previous invalid status.

[the remark this answers is not in the document]
Comment (liutongxuan) [reply src/scheduler/speculative_scheduler.cpp:109]: Actually, better solution could be a standalone class like SpecSequence to take responsibility of all spec related sequence functions. But currently have to make a big change to support subclass of Sequence.

File: src/scheduler/speculative_scheduler.cpp (hunk 6 above)
Comment (liutongxuan) [reply src/scheduler/speculative_scheduler.cpp:32]: i suppose only the ssm's output token ids are shared, do we have to use shared block manager to control it?

File: src/scheduler/scheduler_config.h (hunk 5 above)
Comment (liutongxuan) [reply src/scheduler/scheduler_config.h:34]: PSA（Priority Scheduling Algorithm）

File: src/engine/utils.cpp (hunk 2 above)
Comment (liutongxuan) [reply src/engine/utils.cpp:218]: yes, we can change this part follow up with new attention kernels.

File: src/scheduler/speculative_scheduler.cpp (hunk 7 above)
Comment (liutongxuan) [reply src/scheduler/speculative_scheduler.cpp:74]: done

File: src/scheduler/speculative_scheduler.cpp (hunk 8 above)
Comment (liutongxuan) [reply src/scheduler/speculative_scheduler.cpp:75]: done

File: src/scheduler/speculative_scheduler.h (hunk 3 above)
Comment (liutongxuan) [reply src/scheduler/speculative_scheduler.h:43]: done

File: src/scheduler/speculative_scheduler.cpp (hunk 4 above)
Comment (liutongxuan) [reply src/scheduler/speculative_scheduler.cpp:29]: done

File: src/engine/engine.cpp (hunk 1 above)
Comment (liutongxuan) [reply src/engine/engine.cpp:328]: good catch here.

**Comment by guocuimi** [comment]
LGTM, thanks for addressing my comments.
